src/main/java/io/reactivex/Scheduler.java

```diff
@@ -106,13 +106,10 @@ public Disposable schedule(Runnable run) {
         }
         
         public Disposable schedulePeriodically(Runnable run, final long initialDelay, final long period, final TimeUnit unit) {
-            final SerialDisposable first = new SerialDisposable();
-
-            final SerialDisposable sd = new SerialDisposable(first);
-            
             final Runnable decoratedRun = RxJavaPlugins.onSchedule(run);
-            
-            first.replace(schedule(new Runnable() {
+
+            final SerialDisposable sd = new SerialDisposable();
```

Comment:
Interesting. All the tests passed so I assumed it was safe. Can you show me
in a test how what that would look like? I don't quite follow how sd is
exposed such that it might be changed.

On Sat, Aug 6, 2016, 6:52 PM David Karnok <notifications@github.com> wrote:

> In src/main/java/io/reactivex/Scheduler.java
> <https://github.com/ReactiveX/RxJava/pull/4290#discussion_r73793170>:
>
> >              final Runnable decoratedRun = RxJavaPlugins.onSchedule(run);
> > -
> > -            first.replace(schedule(new Runnable() {
> > +
> > +            final SerialDisposable sd = new SerialDisposable();
>
> You can combine the two into a single class but it will cost 2
> AtomicReferences either way:
>
> class TwoDisposable extends AtomicReference<Disposable> implements Disposable {
>    final AtomicReference<Disposable> first;
>
>    @Override
>    public void replaceFirst(Disposable d) {
>        DisposableHelper.replace(first, d);
>    }
>
>    public void replaceSecond(Disposable d) {
>        DisposableHelper.replace(this, d);
>    }
>
>    @Override
>    public void dispose() {
>        DisposableHelper.dispose(this);
>        DisposableHelper.dispose(second);
>    }
> }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ReactiveX/RxJava/pull/4290/files/9c4682ba91f7fc56681676abc14dc17fbdb40534#r73793170>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEEEc_yLhwZgdsMEOZuD9EmgEcbHJE4ks5qdRAcgaJpZM4JeYXU>
> .
>